### ISSUE TITLE
Declaring constants before structs is a parser error

### ISSUE THREAD

**Issue by axic** [Zokrates/ZoKrates#851]
```
const u32 A = 0

struct State {
	u32 x
}
```

```
Compiling const.zok

Compilation failed:

const.zok: --> 3:1
  |
3 | struct State {␊
  | ^---
  |
  = expected EOI, function_definition, or const_definition
```

Is there a reason for this or only to discourage people trying to use constant as array sizes in structs?

I see the pest file has a very strict ordering:
```
file = { SOI ~ NEWLINE* ~ pragma? ~ NEWLINE* ~ import_directive* ~ NEWLINE* ~ ty_struct_definition* ~ NEWLINE* ~ const_definition* ~ NEWLINE* ~ function_definition* ~ EOI }
```

**Comment by dark64** [comment]
Hello @axic, the ordering is very strict right now yes. We are working on easing it a bit. Using constants as array sizes will be possible soon, but is not supported yet.

**Comment by axic** [comment]
Thanks!

As as I side note: I did a quick change in pest just to see how many things break, but managed to trigger a rustc ICE.

```diff
diff --git a/zokrates_parser/src/zokrates.pest b/zokrates_parser/src/zokrates.pest
index 8280154e..bfcc2a63 100644
--- a/zokrates_parser/src/zokrates.pest
+++ b/zokrates_parser/src/zokrates.pest
@@ -1,9 +1,11 @@
 
-file = { SOI ~ NEWLINE* ~ pragma? ~ NEWLINE* ~ import_directive* ~ NEWLINE* ~ ty_struct_definition* ~ NEWLINE* ~ const_definition* ~ NEWLINE* ~ function_definition* ~ EOI }
+file = { SOI ~ NEWLINE* ~ pragma? ~ NEWLINE* ~ import_directive* ~ NEWLINE* ~ const_or_struct_definition* ~ NEWLINE* ~ function_definition* ~ EOI }
 
 pragma = { "#pragma" ~ "curve" ~ curve }
 curve = @{ (ASCII_ALPHANUMERIC | "_") * }
 
+const_or_struct_definition = { ty_struct_definition | const_definition }
+
 import_directive = { main_import_directive | from_import_directive }
 from_import_directive = { "from" ~ "\"" ~ import_source ~ "\"" ~ "import" ~ import_symbol_list ~ NEWLINE* }
 main_import_directive = { "import" ~ "\"" ~ import_source ~ "\"" ~ ("as" ~ identifier)? ~ NEWLINE+ }
```

rustc:
```
$ cargo build
   Compiling zokrates_parser v0.2.1 (/Users/alex/Projects/anotherzkevm/ZoKrates-src/zokrates_parser)
   Compiling zokrates_pest_ast v0.2.1 (/Users/alex/Projects/anotherzkevm/ZoKrates-src/zokrates_pest_ast)
   Compiling zokrates_core v0.6.1 (/Users/alex/Projects/anotherzkevm/ZoKrates-src/zokrates_core)
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `Some(Fingerprint(7459184979060275217, 10983878263251645031))`,
 right: `Some(Fingerprint(7676583466855060205, 7169415237273974693))`: found unstable fingerprints for predicates_of(zokrates_common[c001]::Resolver): GenericPredicates { parent: None, predicates: [(Binder(TraitPredicate(<E as std::marker::Sized>), []), /Users/alex/Projects/anotherzkevm/ZoKrates/zokrates_common/src/lib.rs:3:20: 3:21 (#0)), (Binder(TraitPredicate(<Self as zokrates_common::Resolver<E>>), []), /Users/alex/Projects/anotherzkevm/ZoKrates/zokrates_common/src/lib.rs:3:1: 9:2 (#0))] }', /rustc/42816d61ead7e46d462df997958ccfd514f8c21c/compiler/rustc_query_system/src/query/plumbing.rs:593:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (42816d61e 2021-04-24) running on x86_64-apple-darwin

note: compiler flags: -C embed-bitcode=no -C split-debuginfo=unpacked -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [predicates_of] computing predicates of `zokrates_common::Resolver`
#1 [typeck] type-checking `imports::<impl at zokrates_core/src/imports.rs:149:1: 378:2>::apply_imports`
end of query stack
error: could not compile `zokrates_core`

To learn more, run the command again with --verbose.
```

**Comment by Schaeff** [comment]
@axic can you stop breaking everything please 😆 

Yeah we'll have constants everywhere soon

**Comment by axic** [comment]
This seems to be fixed in 0.7.2 by #863?

**Comment by Schaeff** [comment]
Yep! Btw @axic had similar issues in Rust because of incremental compilation - usually clearing your target folder helps
